#!/usr/bin/env python
PACKAGE = "fiss_planner"

from dynamic_reconfigure.parameter_generator_catkin import *

gen = ParameterGenerator()

gen.add("check_collision", bool_t, 0, "Default: True", True)
gen.add("use_async", bool_t, 0, "Default: True", True)
gen.add("use_heuristic", bool_t, 0, "Default: True", True)
gen.add("tick_t", double_t, 0, "Default: 0.2", 0.2, 0, 1.0)

gen.add("curr_lane_width", double_t, 0, "Default: 5.0", 5.0, 0, 10.0)
gen.add("left_lane_width", double_t, 0, "Default: 3.0", 3.0, 0, 5.0)
gen.add("right_lane_width", double_t, 0, "Default: 3.0", 3.0, 0, 5.0)
gen.add("center_offset", double_t, 0, "Default: 0", 0, -1.0, 1.0)
gen.add("num_width", int_t, 0, "Default: 9", 9, 1, 15)

gen.add("highest_speed", double_t, 0, "Default: 40.0[km/h]", 40.0, 1.0, 100.0)
gen.add("lowest_speed", double_t, 0, "Default: 20.0[km/h]", 20.0, 1.0, 100.0)
gen.add("num_speed", int_t, 0, "Default: 9", 9, 1, 15)

gen.add("max_t", double_t, 0, "Default: 5.0", 5.0, 0, 10.0)
gen.add("min_t", double_t, 0, "Default: 3.0", 3.0, 0, 10.0)
gen.add("num_t", int_t, 0, "Default: 5", 5, 1, 15)

gen.add("k_heuristic", double_t, 0, "Default: 1.0", 1.0, 0, 10.0)
gen.add("k_diff", double_t, 0, "Default: 1.0", 1.0, 0, 1.0)
gen.add("k_time", double_t, 0, "Default: 1.0", 1.0, 0, 1.0)
gen.add("k_jerk", double_t, 0, "Default: 1.0", 1.0, 0, 1.0)
gen.add("k_lat", double_t, 0, "Default: 1.0", 1.0, 0, 1.0)
gen.add("k_lon", double_t, 0, "Default: 1.0", 1.0, 0, 1.0)
# gen.add("k_obstacle", double_t, 0, "Default: 0.5", 0.5, 0, 1.0)

gen.add("max_speed", double_t, 0, "Default: 120.0 [km/h]", 120.0, 0.0, 160.0)
gen.add("max_acceleration", double_t, 0, "Default: 3.0 [m/s^2]", 3.0, 0.0, 5.0)
gen.add("max_deceleration", double_t, 0, "Default: 3.0 [m/s^2]", 3.0, 0.0, 5.0)
gen.add("max_curvature", double_t, 0, "Default: 3.0 [rad/m]", 3.0, 0.0, 5.0)
gen.add("max_jerk_lon", double_t, 0, "Default: 1.0 [m/s^3]", 1.0, 0.0, 5.0)
gen.add("max_jerk_lat", double_t, 0, "Default: 1.0 [m/s^3]", 1.0, 0.0, 5.0)

gen.add("vehicle_length", double_t, 0, "Default: 5.0 [m]", 5.0, 0.0, 10.0)
gen.add("vehicle_width", double_t, 0, "Default: 2.0 [m]", 2.0, 0.0, 5.0)
gen.add("safety_margin_lon", double_t, 0, "Default: 1.0 [ratio]", 1.0, 0.0, 3.0)
gen.add("safety_margin_lat", double_t, 0, "Default: 0.5 [ratio]", 0.5, 0.0, 1.0)
gen.add("safety_margin_soft", double_t, 0, "Default: 0.5 [ratio]", 0.5, 0.0, 3.0)

gen.add("PID_Kp", double_t, 1, "Default: 0.15", 0.15, 0, 1.0)
gen.add("PID_Ki", double_t, 1, "Default: 0.01", 0.01, 0, 1.0)
gen.add("PID_Kd", double_t, 1, "Default: 0.0", 0.0, 0, 1.0)
gen.add("num_wp_look_ahead", int_t, 1, "Default: 2", 2, 0, 5)
gen.add("stanley_overall_gain", double_t, 1, "Default: 0.6", 0.6, 0, 1.0)
gen.add("track_error_gain", double_t, 1, "Default: 1.0", 1.0, 0, 1.0)

gen.add("traj_max_size", int_t, 0, "Default: 10", 10, 5, 30)
gen.add("traj_min_size", int_t, 0, "Default: 5", 5, 5, 20)

exit(gen.generate(PACKAGE, "fiss_planner", "fiss_planner_"))
